coap-core/src/main/java/com/mbed/coap/utils/CacheImpl.java

```diff
@@ -0,0 +1,124 @@
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CacheImpl<K extends ExpiringKey, V> implements Cache<K, V> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DuplicationDetector.class.getName());
+    private final Lock REDUCE_LOCK = new ReentrantLock();
+    private final ConcurrentHashMap<K, V> underlying;
+    private final long maxSize;
+    private final long overSizeMargin;
+    private final long warnIntervalMillis;
+    private long cleanIntervalMillis;
+    private long nextWarnMessage;
+    private final String cacheName;
+    private final ScheduledExecutorService scheduledExecutor;
+    private ScheduledFuture<?> cleanWorkerFut;
+
+    public CacheImpl(String cacheName, long maxSize, long cleanIntervalMillis, long warnIntervalMillis, ScheduledExecutorService scheduledExecutor) {
+        this.cacheName = cacheName;
+        this.maxSize = maxSize;
+        this.cleanIntervalMillis = cleanIntervalMillis;
+        this.warnIntervalMillis = warnIntervalMillis;
+        this.scheduledExecutor = scheduledExecutor;
+        this.overSizeMargin = maxSize / 100; //1%
+        underlying = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void setCleanIntervalMillis(long cleanIntervalMillis) {
+        this.cleanIntervalMillis = cleanIntervalMillis;
+    }
+
+    @Override
+    public void start() {
+        cleanWorkerFut = scheduledExecutor.scheduleWithFixedDelay(() -> clean(), cleanIntervalMillis, cleanIntervalMillis, TimeUnit.MILLISECONDS);
+    }
+
+    @Override
+    public void stop() {
+        cleanWorkerFut.cancel(true);
+    }
+
+    @Override
+    public V putIfAbsent(K key, V value) {
+        return underlying.putIfAbsent(key, value);
+    }
+
+    @Override
+    public void put(K key, V value) {
+        underlying.put(key, value);
+    }
+
+    @Override
+    public V get(K key) {
+        return underlying.get(key);
+    }
+
+    @Override
+    public void clean() {
+        int removedItems = 0;
+        Iterator<K> it = underlying.keySet().iterator();
+        final long currentTimeMillis = System.currentTimeMillis();
+        while (it.hasNext()) {
+            if (!it.next().isValid(currentTimeMillis)) {
+                it.remove();
+                removedItems++;
+            }
+        }
+        if (LOGGER.isTraceEnabled() && removedItems > 0) {
+            LOGGER.trace("CoAP request duplicate list, non valid items removed: " + removedItems + " ");
+        }
+    }
+
+    @Override
+    public void cleanupBulk() {
+        if (underlying.size() > maxSize + overSizeMargin && REDUCE_LOCK.tryLock()) {
+            try {
+                //reduce map size in bulk
+                Iterator<K> it = underlying.keySet().iterator();
+                try {
+                    for (int i = 0; i <= overSizeMargin && it.hasNext(); i++) {
```

Comment:
This removes the entries randomly (by hash). Have you considered using LinkedHashMap()?
Do we still need this feature to remove overSizeMargin at once?

Reply:
Good catch! This is something that has been there. I'm thinking that it's out of scope of this change. However, I agree that this is a bit problematic.

Reply:
The cache injected by the coap-service works differently, I suppose.

Reply:
Yes it will.